Library/MeterGeometry.cpp

```diff
@@ -0,0 +1,370 @@
+							if (!ReplaceShapeOption(m_Shapes[shape.first], option.first.c_str(), optionValue.c_str()))
+								continue;
+						}
+						else
+							if (!ParseShape(m_Shapes[shape.first], option.first.c_str(), optionValue.c_str()))
+								continue;
+				}
+			}
+			return true;
+		}
+		else if (m_NeedsRedraw)
+		{
+			m_NeedsRedraw = false;
+			return true;
+		}
+	}
+	return false;
+}
+
+bool MeterGeometry::Draw(Gfx::Canvas & canvas)
+{
+	if (!Meter::Draw(canvas)) return false;
+	
+	for (const auto& shape : m_Shapes) {
+		if (shape.second.m_Shape) {
+			D2D1_POINT_2F centerPoint;
+			centerPoint = D2D1::Point2F(shape.second.m_RotationCenter.x, shape.second.m_RotationCenter.y);
+			D2D1_MATRIX_3X2_F transform = D2D1::Matrix3x2F(
+				D2D1::Matrix3x2F::Rotation(shape.second.m_Rotation, centerPoint) *
+
+				D2D1::Matrix3x2F::Skew(shape.second.m_Skew.x, shape.second.m_Skew.y, D2D1::Point2F(shape.second.m_Bounds.left, shape.second.m_Bounds.top)) *
+				D2D1::Matrix3x2F::Translation(shape.second.m_Offset) * D2D1::Matrix3x2F::Translation(D2D1::SizeF(GetMeterRectPadding().X, GetMeterRectPadding().Y)) *
+				D2D1::Matrix3x2F::Scale(shape.second.m_Scale, D2D1::Point2F(shape.second.m_Bounds.left, shape.second.m_Bounds.top))
+				);
+			canvas.DrawGeometry(shape.second, transform);
+		}
+	}
+	return true;
+}
+/*
+** Not necessary for the Geometry meter to be bound to any measures
+**
+*/
+void MeterGeometry::BindMeasures(ConfigParser & parser, const WCHAR * section)
+{
+	if (BindPrimaryMeasure(parser, section, true))
+	{
+		BindSecondaryMeasures(parser, section);
+	}
+}
+
+Microsoft::WRL::ComPtr<ID2D1Geometry> MeterGeometry::ParseRect(GeometryShape& shape, RECT& rect)
+{
+	D2D1_RECT_F geo_rect = D2D1::RectF(rect.left, rect.top, rect.right + rect.left, rect.bottom + rect.top);
+	geo_rect.left += shape.m_OutlineWidth / 4;
+	geo_rect.top += shape.m_OutlineWidth / 4;
+	return Gfx::Canvas::CreateRectangle(geo_rect);
+}
+
+double MeterGeometry::ParseRotation(const WCHAR* string, double defaultValue, GeometryShape& shape)
+{
+	std::vector<std::wstring> tokens = CustomTokenize(string, L",");
+	double rotation = 0;
+	if (tokens.size() > 0)  rotation = ConfigParser::ParseDouble(tokens[0].c_str(), defaultValue);
+	if (tokens.size() > 2) {
+		shape.m_RotationCenter = ParsePoint((tokens[1] + L"," + tokens[2]).c_str(), FLT_MIN);
+		if (shape.m_RotationCenter.x == FLT_MIN) shape.m_RotationCenter.x = (shape.m_Bounds.right - shape.m_Bounds.left) / 2 - shape.m_Bounds.left;
+		if (shape.m_RotationCenter.y == FLT_MIN) shape.m_RotationCenter.y = (shape.m_Bounds.bottom - shape.m_Bounds.top) / 2 - shape.m_Bounds.top;
+	}
+	else {
+		shape.m_RotationCenter.x = (shape.m_Bounds.right - shape.m_Bounds.left) / 2 - shape.m_Bounds.left;
+		shape.m_RotationCenter.y = (shape.m_Bounds.bottom - shape.m_Bounds.top) / 2 - shape.m_Bounds.top;
+	}
+	return rotation;
+}
+
+std::vector<std::wstring> MeterGeometry::CustomTokenize(const std::wstring& str, const std::wstring& delimiters)
```

Comment:
Can you explain how why we need this (i.e. what does this do differently compared to `ConfigParser::Tokenize`)?